File: experimental/python/gui/statusbar_button.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
====================================
|module_summary| statusbar_button.py
====================================

TOWRITE


Classes summary:
================

============================ ============================
:class:`~StatusBarButton`    A `QToolButton`_ GUI element to be injected into a `QStatusBar`_.
============================ ============================

---------------------------------------------------------
"""

#-Imports.---------------------------------------------------------------------
#--Python Imports.
import os

#--PySide/PyQt Imports.
if PYSIDE:
    from PySide.QtCore import qDebug, Qt, Slot
    from PySide.QtGui import QApplication, QAction, QIcon, QMenu, QToolButton
elif PYQT4:
    import sip
    sip.setapi('QString', 2)
    sip.setapi('QVariant', 2)
    from PyQt4.QtCore import qDebug, Qt
    from PyQt4.QtCore import pyqtSlot as Slot
    from PyQt4.QtGui import QApplication, QAction, QIcon, QMenu, QToolButton


class StatusBarButton(QToolButton):
    """
    Subclass of `QToolButton`_

    A `QToolButton`_ GUI element to be injected into a `QStatusBar`_.

    .. sphinx_generate_methods_summary::
       StatusBarButton
    """
    def __init__(self, buttonText, mw, statbar, parent=None):
        """
        Default class constructor.

        :param `buttonText`: the text for the button label.
        :type `buttonText`: QString
        :param `mw`: The application's MainWindow instance.
        :type `mw`: `QMainWindow`_
        :param `statbar`: The parent statusbar instance of this button.
        :type `statbar`: `QStatusBar`_
        :param `parent`: Pointer to a parent widget instance.
        :type `parent`: `QWidget`_
        """
        super(StatusBarButton, self).__init__(parent)

        self.mainWin = mw
        self.statusbar = statbar

        self.setObjectName("StatusBarButton" + buttonText)

        self.setText(buttonText)
        self.setAutoRaise(True)
        self.setCheckable(True)

        objName = self.objectName()
        if   objName == "StatusBarButtonSNAP":
            self.toggled.connect(self.toggleSnap)
        elif objName == "StatusBarButtonGRID":
            self.toggled.connect(self.toggleGrid)
        elif objName == "StatusBarButtonRULER":
            self.toggled.connect(self.toggleRuler)
        elif objName == "StatusBarButtonORTHO":
            self.toggled.connect(self.toggleOrtho)
        elif objName == "StatusBarButtonPOLAR":
            self.toggled.connect(self.togglePolar)
        elif objName == "StatusBarButtonQSNAP":
            self.toggled.connect(self.toggleQSnap)
        elif objName == "StatusBarButtonQTRACK":
            self.toggled.connect(self.toggleQTrack)
        elif objName == "StatusBarButtonLWT":
            self.toggled.connect(self.toggleLwt)

    def contextMenuEvent(self, event):
        """
        Handles the ``contextMenuEvent`` event for :class:`StatusBarButton`.

        :param `event`: A `QContextMenuEvent`_ to be processed.
        """
        QApplication.setOverrideCursor(Qt.ArrowCursor)
        menu = QMenu(self)
        objName = self.objectName()
        themeDir = self.mainWin.gIconDir + os.sep + self.mainWin.getSettingsGeneralIconTheme() + os.sep
        s = "&Settings..."
        if objName == "StatusBarButtonSNAP":
            settingsSnapAction = QAction(QIcon(themeDir + "gridsnapsettings.png"), s, menu)
            settingsSnapAction.triggered.connect(self.settingsSnap)
            menu.addAction(settingsSnapAction)

        elif objName == "StatusBarButtonGRID":
            settingsGridAction = QAction(QIcon(themeDir + "gridsettings.png"), s, menu)
            settingsGridAction.triggered.connect(self.settingsGrid)
            menu.addAction(settingsGridAction)

        elif objName == "StatusBarButtonRULER":
            settingsRulerAction = QAction(QIcon(themeDir + "rulersettings.png"), s, menu)
            settingsRulerAction.triggered.connect(self.settingsRuler)
            menu.addAction(settingsRulerAction)

        elif objName == "StatusBarButtonORTHO":
            settingsOrthoAction = QAction(QIcon(themeDir + "orthosettings.png"), s, menu)
            settingsOrthoAction.triggered.connect(self.settingsOrtho)
            menu.addAction(settingsOrthoAction)

        elif objName == "StatusBarButtonPOLAR":
            settingsPolarAction = QAction(QIcon(themeDir + "polarsettings.png"), s, menu)
            settingsPolarAction.triggered.connect(self.settingsPolar)
            menu.addAction(settingsPolarAction)

        elif objName == "StatusBarButtonQSNAP":
            settingsQSnapAction = QAction(QIcon(themeDir + "qsnapsettings.png"), s, menu)
            settingsQSnapAction.triggered.connect(self.settingsQSnap)
            menu.addAction(settingsQSnapAction)

        elif objName == "StatusBarButtonQTRACK":
            settingsQTrackAction = QAction(QIcon(themeDir + "qtracksettings.png"), s, menu)
            settingsQTrackAction.triggered.connect(self.settingsQTrack)
            menu.addAction(settingsQTrackAction)

        elif objName == "StatusBarButtonLWT":
            gview = self.mainWin.activeView()
            if gview:

                enableRealAction = QAction(QIcon(themeDir + "realrender.png"), "&RealRender On", menu)
                enableRealAction.setEnabled(not gview.isRealEnabled())
                enableRealAction.triggered.connect(self.enableReal)
                menu.addAction(enableRealAction)

                disableRealAction = QAction(QIcon(themeDir + "realrender.png"), "&RealRender Off", menu)
                disableRealAction.setEnabled(gview.isRealEnabled())
                disableRealAction.triggered.connect(self.disableReal)
                menu.addAction(disableRealAction)

            settingsLwtAction = QAction(QIcon(themeDir + "lineweightsettings" + ".png"), s, menu)
            settingsLwtAction.triggered.connect(self.settingsLwt)
            menu.addAction(settingsLwtAction)

        menu.exec_(event.globalPos())
        QApplication.restoreOverrideCursor()
        self.statusbar.clearMessage()

    # Slots ------------------------------------------------------------------

    @Slot()
    def settingsSnap(self):
        """
        Open the `Settings_Dialog` on the "Snap" notebook tab page.
        """
        self.mainWin.settingsDialog("Snap")

    @Slot()
    def settingsGrid(self):
        """
        Open the `Settings_Dialog` on the "Grid/Ruler" notebook tab page.
        """
        self.mainWin.settingsDialog("Grid/Ruler")

    @Slot()
    def settingsRuler(self):
        """
        Open the `Settings_Dialog` on the "Grid/Ruler" notebook tab page.
        """
        self.mainWin.settingsDialog("Grid/Ruler")

    @Slot()
    def settingsOrtho(self):
        """
        Open the `Settings_Dialog` on the "Ortho/Polar" notebook tab page.
        """
        self.mainWin.settingsDialog("Ortho/Polar")

    @Slot()
    def settingsPolar(self):
        """
        Open the `Settings_Dialog` on the "Ortho/Polar" notebook tab page.
        """
        self.mainWin.settingsDialog("Ortho/Polar")

    @Slot()
    def settingsQSnap(self):
        """
        Open the `Settings_Dialog` on the "QuickSnap" notebook tab page.
        """
        self.mainWin.settingsDialog("QuickSnap")

    @Slot()
    def settingsQTrack(self):
        """
        Open the `Settings_Dialog` on the "QuickTrack" notebook tab page.
        """
        self.mainWin.settingsDialog("QuickTrack")

    @Slot()
    def settingsLwt(self):
        """
        Open the `Settings_Dialog` on the "LineWeight" notebook tab page.
        """
        self.mainWin.settingsDialog("LineWeight")

    @Slot(bool)
    def toggleSnap(self, on):
        """
        TOWRITE

        :param `on`: TOWRITE
        :type `on`: bool
        """
        qDebug("StatusBarButton toggleSnap()")
        gview = self.mainWin.activeView()
        if gview:
            gview.toggleSnap(on)

    @Slot(bool)
    def toggleGrid(self, on):
        """
        TOWRITE

        :param `on`: TOWRITE
        :type `on`: bool
        """
        qDebug("StatusBarButton toggleGrid()")
        gview = self.mainWin.activeView()
        if gview:
            gview.toggleGrid(on)

    @Slot(bool)
    def toggleRuler(self, on):
        """
        TOWRITE

        :param `on`: TOWRITE
        :type `on`: bool
        """
        qDebug("StatusBarButton toggleRuler()")
        gview = self.mainWin.activeView()
        if gview:
            gview.toggleRuler(on)

    @Slot(bool)
    def toggleOrtho(self, on):
        """
        TOWRITE

        :param `on`: TOWRITE
        :type `on`: bool
        """
        qDebug("StatusBarButton toggleOrtho()")
        gview = self.mainWin.activeView()
        if gview:
            gview.toggleOrtho(on)

    @Slot(bool)
    def togglePolar(self, on):
        """
        TOWRITE

        :param `on`: TOWRITE
        :type `on`: bool
        """
        qDebug("StatusBarButton togglePolar()")
        gview = self.mainWin.activeView()
        if gview:
            gview.togglePolar(on)

    @Slot(bool)
    def toggleQSnap(self, on):
        """
        TOWRITE

        :param `on`: TOWRITE
        :type `on`: bool
        """
        qDebug("StatusBarButton toggleQSnap()")
        gview = self.mainWin.activeView()
        if gview:
            gview.toggleQSnap(on)

    @Slot(bool)
    def toggleQTrack(self, on):
        """
        TOWRITE

        :param `on`: TOWRITE
        :type `on`: bool
        """
        qDebug("StatusBarButton toggleQTrack()")
        gview = self.mainWin.activeView()
        if gview:
            gview.toggleQTrack(on)

    @Slot(bool)
    def toggleLwt(self, on):
        """
        TOWRITE

        :param `on`: TOWRITE
        :type `on`: bool
        """
        qDebug("StatusBarButton toggleLwt()")
        gview = self.mainWin.activeView()
        if gview:
            gview.toggleLwt(on)

    @Slot()
    def enableLwt(self):
        """"""
        qDebug("StatusBarButton enableLwt()")
        gview = self.mainWin.activeView()
        if gview:
            if not gview.isLwtEnabled():
                gview.toggleLwt(True)

    @Slot()
    def disableLwt(self):
        """"""
        qDebug("StatusBarButton disableLwt()")
        gview = self.mainWin.activeView()
        if gview:
            if gview.isLwtEnabled():
                gview.toggleLwt(False)

    @Slot()
    def enableReal(self):
        """"""
        qDebug("StatusBarButton enableReal()")
        gview = self.mainWin.activeView()
        if gview:
            gview.toggleReal(True)

    @Slot()
    def disableReal(self):
        """"""
        qDebug("StatusBarButton disableReal()")
        gview = self.mainWin.activeView()
        if gview:
            gview.toggleReal(False)


# kate: bom off; indent-mode python; indent-width 4; replace-trailing-space-save on;

File: embroidermodder2/statusbar-button.cpp
/**
 *  Embroidermodder 2.
 *
 *  ------------------------------------------------------------
 *
 *  Copyright 2013-2022 The Embroidermodder Team
 *  Embroidermodder 2 is Open Source Software.
 *  See LICENSE for licensing terms.
 *
 *  ------------------------------------------------------------
 *
 *  Use Python's PEP7 style guide.
 *      https://peps.python.org/pep-0007/
 */

#include "embroidermodder.h"

/**
 * .
 */
StatusBarButton::StatusBarButton(QString buttonText, StatusBar* statbar, QWidget *parent) : QToolButton(parent)
{
    statusbar = statbar;

    this->setObjectName("StatusBarButton" + buttonText);

    this->setText(buttonText);
    this->setAutoRaise(true);
    this->setCheckable(true);

    if (objectName() == "StatusBarButtonSNAP") {
        connect(this, SIGNAL(toggled(bool)), this, SLOT(toggleSnap(bool))); 
    }
    else if (objectName() == "StatusBarButtonGRID") {
        connect(this, SIGNAL(toggled(bool)), this, SLOT(toggleGrid(bool)));
    }
    else if (objectName() == "StatusBarButtonRULER") {
        connect(this, SIGNAL(toggled(bool)), this, SLOT(toggleRuler(bool)));
    }
    else if (objectName() == "StatusBarButtonORTHO") {
        connect(this, SIGNAL(toggled(bool)), this, SLOT(toggleOrtho(bool)));
    }
    else if (objectName() == "StatusBarButtonPOLAR") {
        connect(this, SIGNAL(toggled(bool)), this, SLOT(togglePolar(bool)));
    }
    else if (objectName() == "StatusBarButtonQSNAP") {
        connect(this, SIGNAL(toggled(bool)), this, SLOT(toggleQSnap(bool)));
    }
    else if (objectName() == "StatusBarButtonQTRACK") {
        connect(this, SIGNAL(toggled(bool)), this, SLOT(toggleQTrack(bool)));
    }
    else if (objectName() == "StatusBarButtonLWT") {
        connect(this, SIGNAL(toggled(bool)), this, SLOT(toggleLwt(bool)));
    }
}

/**
 * .
 */
void
StatusBarButton::contextMenuEvent(QContextMenuEvent *event)
{
    QApplication::setOverrideCursor(Qt::ArrowCursor);
    QMenu menu(this);
    if (objectName() == "StatusBarButtonSNAP") {
        QAction* settingsSnapAction = new QAction(_mainWin->create_icon("gridsnapsettings"), "&Settings...", &menu);
        connect(settingsSnapAction, SIGNAL(triggered()), this, SLOT(settingsSnap()));
        menu.addAction(settingsSnapAction);
    }
    else if (objectName() == "StatusBarButtonGRID") {
        QAction* settingsGridAction = new QAction(_mainWin->create_icon("gridsettings"), "&Settings...", &menu);
        connect(settingsGridAction, SIGNAL(triggered()), this, SLOT(settingsGrid()));
        menu.addAction(settingsGridAction);
    }
    else if (objectName() == "StatusBarButtonRULER") {
        QAction* settingsRulerAction = new QAction(_mainWin->create_icon("rulersettings"), "&Settings...", &menu);
        connect(settingsRulerAction, SIGNAL(triggered()), this, SLOT(settingsRuler()));
        menu.addAction(settingsRulerAction);
    }
    else if (objectName() == "StatusBarButtonORTHO") {
        QAction* settingsOrthoAction = new QAction(_mainWin->create_icon("orthosettings"), "&Settings...", &menu);
        connect(settingsOrthoAction, SIGNAL(triggered()), this, SLOT(settingsOrtho()));
        menu.addAction(settingsOrthoAction);
    }
    else if (objectName() == "StatusBarButtonPOLAR") {
        QAction* settingsPolarAction = new QAction(_mainWin->create_icon("polarsettings"), "&Settings...", &menu);
        connect(settingsPolarAction, SIGNAL(triggered()), this, SLOT(settingsPolar()));
        menu.addAction(settingsPolarAction);
    }
    else if (objectName() == "StatusBarButtonQSNAP") {
        QAction* settingsQSnapAction = new QAction(_mainWin->create_icon("qsnapsettings"), "&Settings...", &menu);
        connect(settingsQSnapAction, SIGNAL(triggered()), this, SLOT(settingsQSnap()));
        menu.addAction(settingsQSnapAction);
    }
    else if (objectName() == "StatusBarButtonQTRACK") {
        QAction* settingsQTrackAction = new QAction(_mainWin->create_icon("qtracksettings"), "&Settings...", &menu);
        connect(settingsQTrackAction, SIGNAL(triggered()), this, SLOT(settingsQTrack()));
        menu.addAction(settingsQTrackAction);
    }
    else if (objectName() == "StatusBarButtonLWT") {
        View* gview = activeView();
        if (gview) {
            QAction* enableRealAction = new QAction(_mainWin->create_icon("realrender"), "&RealRender On", &menu);
            enableRealAction->setEnabled(!gview->isRealEnabled());
            connect(enableRealAction, SIGNAL(triggered()), this, SLOT(enableReal()));
            menu.addAction(enableRealAction);

            QAction* disableRealAction = new QAction(_mainWin->create_icon("realrender"), "&RealRender Off", &menu);
            disableRealAction->setEnabled(gview->isRealEnabled());
            connect(disableRealAction, SIGNAL(triggered()), this, SLOT(disableReal()));
            menu.addAction(disableRealAction);
        }

        QAction* settingsLwtAction = new QAction(_mainWin->create_icon("lineweightsettings"), "&Settings...", &menu);
        connect(settingsLwtAction, SIGNAL(triggered()), this, SLOT(settingsLwt()));
        menu.addAction(settingsLwtAction);
    }
    menu.exec(event->globalPos());
    QApplication::restoreOverrideCursor();
    statusbar->clearMessage();
}

void StatusBarButton::settingsSnap()
{
    _mainWin->settingsDialog("Snap");
}

void StatusBarButton::settingsGrid()
{
    _mainWin->settingsDialog("Grid/Ruler");
}

void StatusBarButton::settingsRuler()
{
    _mainWin->settingsDialog("Grid/Ruler");
}

void StatusBarButton::settingsOrtho()
{
    _mainWin->settingsDialog("Ortho/Polar");
}

void StatusBarButton::settingsPolar()
{
    _mainWin->settingsDialog("Ortho/Polar");
}

void StatusBarButton::settingsQSnap()
{
    _mainWin->settingsDialog("QuickSnap");
}

void StatusBarButton::settingsQTrack()
{
    _mainWin->settingsDialog("QuickTrack");
}

void StatusBarButton::settingsLwt()
{
    _mainWin->settingsDialog("LineWeight");
}

void StatusBarButton::toggleSnap(bool on)
{
    debug_message("StatusBarButton toggleSnap()");
    View* gview = activeView();
    if (gview) {
        gview->toggleSnap(on);
    }
}

void StatusBarButton::toggleGrid(bool on)
{
    debug_message("StatusBarButton toggleGrid()");
    View* gview = activeView();
    if (gview) {
        gview->toggleGrid(on);
    }
}

void StatusBarButton::toggleRuler(bool on)
{
    debug_message("StatusBarButton toggleRuler()");
    View* gview = activeView();
    if (gview) {
        gview->toggleRuler(on);
    }
}

void StatusBarButton::toggleOrtho(bool on)
{
    debug_message("StatusBarButton toggleOrtho()");
    View* gview = activeView();
    if (gview) {
        gview->toggleOrtho(on);
    }
}

void StatusBarButton::togglePolar(bool on)
{
    debug_message("StatusBarButton togglePolar()");
    View* gview = activeView();
    if (gview) {
        gview->togglePolar(on);
    }
}

void StatusBarButton::toggleQSnap(bool on)
{
    debug_message("StatusBarButton toggleQSnap()");
    View* gview = activeView();
    if (gview) {
        gview->toggleQSnap(on);
    }
}

void StatusBarButton::toggleQTrack(bool on)
{
    debug_message("StatusBarButton toggleQTrack()");
    View* gview = activeView();
    if (gview) {
        gview->toggleQTrack(on);
    }
}

void StatusBarButton::toggleLwt(bool on)
{
    debug_message("StatusBarButton toggleLwt()");
    View* gview = activeView();
    if (gview) {
        gview->toggleLwt(on);
    }
}

void StatusBarButton::enableLwt()
{
    debug_message("StatusBarButton enableLwt()");
    View* gview = activeView();
    if (gview) {
        gview->toggleLwt(true);
    }
}

void StatusBarButton::disableLwt()
{
    debug_message("StatusBarButton disableLwt()");
    View* gview = activeView();
    if (gview) {
        gview->toggleLwt(false);
    }
}

void StatusBarButton::enableReal()
{
    debug_message("StatusBarButton enableReal()");
    View* gview = activeView();
    if (gview) {
        gview->toggleReal(true);
    }
}

void StatusBarButton::disableReal()
{
    debug_message("StatusBarButton disableReal()");
    View* gview = activeView();
    if (gview) {
        gview->toggleReal(false);
    }
}
